feature: GOAI Urist port v2

scrdest opened the pull request:
<!-- 
Do not forget to add a changelog when you have made admin/player facing changes that can alter gameplay.
Examples which require a changelog entry include:
* Adding/removing objects that players may interact with, or the way they function.
* Adding/removing/altering admin tools.
* Changing the map.

Examples were changelog entries are optional/not typically required:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

You'll find a README and example file in .\html\changelogs\ for further instructions.

You can also find a template for adding your changelog directly to the PR description here: https://github.com/Baystation12/Baystation12/wiki/Automatic-changelog-generation
-->

Yep. This again. And a cool 10k+ SLOCs to go through!

Featuring an entirely re-done architecture - the heart of the AI is now a Utility AI, blatantly lifting the core design from Infinite Axis Utility System by Dave Mark & Co., with improved and optimized GOAP Planner AI as an optional sidecar for dealing with complex environmental problem-solving (e.g. GODDAMN DOORS REEE).

### Fast and smart

A major consequence of that is that this iteration of GOAI is at least as smart as the previous (with a strong argument that it's much more so), but also _very, very **fast**_. And that's _before_ AI LODs or any other future optimizations.

### Data-driven

The redesign also showcases my progressing conversion to the church of Data-Driven Design. As a result, the AI system is really two things: an **AI Engine** in DM code (because I am a masochist) and an **AI Specification Language** in the form of a bunch of JSON schemas.

The AI Engine provides a pile of abstractions and building blocks (curves, Considerations, ContextFetchers...) that can be used in the JSONs to build actual AIs and SmartObjects. 

Thanks to that, the AI Specs can be edited without an in-depth knowledge of DM's weird quirks. You don't have to be a full-on coder to make GOAI AIs.

They can also be hot-swapped at runtime without recompiling the whole game as long as you're not extending the AI Engine itself. Should make iteration speed muuuuuuuch faster.

### Modular and extentable

Unlike _Certain Other SS13 AI Implementations_, GOAI provides only a cheap, bare minimum shared framework for any gameobject that opts into having GOAI enabled at all.

Everything else - all behaviors and checks - only runs if an AI Agent uses them for something and they run in whatever sequences they are acquired. The only hardcoded state-machines are either the AI Brain's core decision/execution loop or the scripts powering the actions (if you choose to write them that way). The AI will transition fluidly between combat, idling, chores, and everything in between based on whatever's the next logical priority.

Actions can be added or removed on the fly - in fact, they already are. 

To avoid duplicating logic across 20 different mob types, GOAI now includes Sims-style SmartObjects, which are AI Specs attached to gameobjects. Agents can harvest new abilities from those objects - for example, a held gun may provide a Shoot Action, and a nearby Door may provide an OpenDoor Action.

Even the GOAP integration is more flexible than traditional implementations - GOAI plan execution is smart enough to handle third-party interference cancelling out the effects of actions earlier in the plan or short-circuiting the requirements to make the ultimate goal available.

Commit:
wip: Initial file-dump into the repo. Commit for tracking deltas on the branch, mainly.

Commit:
wip: Fixing conflicts

Commit:
Updated to 13.03.2024 GOAI lib

Commit:
chore: Moved everything under urist/ again to get it to compile correctly.

Commit:
fix: Gave '/mob/living/carbon/human's a SmartObject actionset - basic AI-enabled proper sim humans work!

Commit:
feature: Gave hostiles a basic ranged mob SmartObject spec - they can now be GOAI-enabled

(the AI similar to human/testmob, but with Shoot being intrinsic rather than sourced from a Gun SO)

Commit:
feature: Better tick-handling (broken up sleep()s), mass-pause, override for SS13 AI, door and table actions for SS13, tons of log cleanup.

Direct port of:
SHA-1: 1a7bffad75bb4be52633ad4101a89bf10e1c0943
from GOAI repo

Commit:
feature: Significant Coverleap improvements, BBox CF and Grenade logic using it, allowed groupby-like mean aggregations, fixed raytrace ignores and other stuff.

Sync to GOAI commit:
SHA-1: 22948c7db2ed5cf98e7ae966f1bae11608c080ac

Commit:
feature: Generating personalities from template files. List-ified Personality (from a dict object) for simplicity.
feature: Templated Personality generation, faked melee for dev purposes, added style affinities to appropriate attacks.

Port of the following GOAI commits (respectively):
SHA-1: e609609c07d9f85011326dc167a697b5c28d6047
SHA-1: a4dbe5178d6fd88085cd71aa48cfd6f3ab68e7a9

Commit:
fix: Missed some files I guess

Commit:
feature: Added a random factor to AI tickrate to stagger them - both more organic and causes less CPU load spikes. Melee hit-by events. Steadfast personality trait (less Composure loss per hit); basic loss raised to compensate. Fixed Consideration values. Filtering dense turfs in bbox (optionally). Reverse raytraces for enemy-safety checks.
fix: Updating/fixing SS13 AIspecs. Ranged SAs can now use their guns correctly!

sync to GOAI SHA-1: 8057d06ec0397e9367d01c9854f3db31b62471cc

Commit:
fix: Big pile o' stuff; fixed melee, fixed some bad raycast logic, fixed cover data gen, moved out-in-the-open checks to Vision, tuned cover/fight better, added gentler fuzz Considerations

Commit:
fix: Some JSON I forgot to port from the other repo

Commit:
feature: Nigh-fully working Z-level movement! Ladders, stairs and holes all handled. Weird paths may pop up on occasion. Ignoring dead mobs. Split neutrality threshold into enemy/friend thresholds. Wrapper to call an arbitrary method. Macro-ized smartobject declarations for less boilerplate. Gated off some more dev objects behind Dev-only defines.

Commit:
feature: 3d movement fully working (aside from ladder contextfetcher being a bit unsound).

Commit:
fix: Fixes to devcode version of 3d movement.

Mostly shouldn't affect SS13 code, but nice to keep 'em synced.

Commit:
fix: Sync to latest GOAI repo; fixed runaway AI tickrates, cleaned up logging with MOAR MACROS.

scrdest commented on maps/using.dm:
Oops, will revert honf

Commit:
chore: First pass at linter fixes. Definitely not the last.

Commit:
chore: Lints pt. 2

Commit:
chore: Lints pt. 3 - hopefully the last of Dreamchecker ones

Commit:
fix: Lints pt. 4: Ironing is delicious Edition

Commit:
chore: Let's see if that helps with the weird indents lint

Commit:
chore: Fixes some densities. Reverts the previous failed attempt to fix indents. Actually 'fixes' the problem.

(this is a dumb context-free AWK check seemingly designed for indents on typepaths; nobody seems to have thought anyone will need to pass in a typepath IN CODE)

scrdest commented on code/modules/multiz/structures.dm:
```diff
@@ -223,7 +228,9 @@
-	return user.Move(T)
+	var/result = user.Move(T)
+	to_world_log("[user] has climbed [target_ladder] to [T] (Result: [result])")
+	return result
```
Oops, dev changes, will revert

scrdest commented on code/modules/multiz/structures.dm:
```diff
@@ -206,10 +206,15 @@
 	return TRUE
 
 /obj/structure/ladder/proc/climbLadder(mob/user, target_ladder, obj/item/I = null)
+	to_world_log("[user] is climbing [target_ladder]")
```
Oops, dev changes, will revert

scrdest commented on genmap.json:
Should not have been committed, will nuke

scrdest commented on code/modules/urist/GOAI/basics/docs/movementCallstack.png:
Stale diagram, will remove

scrdest commented on code/modules/urist/GOAI/basics/docs/movementFinderLogic.png:
Stale diagram, will remove

scrdest commented on code/modules/urist/GOAI/__ss13/_ss13macros.dm:
DEVJUNK: Feel free to ignore this. This is a port of SS13 code to the Dev repo. It's only here to make syncing between repos easier (can do copy/paste without unpicking individual files).

scrdest commented on code/modules/urist/GOAI/__ss13/_ss13stubmacros.dm:
DEVJUNK: Feel free to ignore this. This is a port of SS13 code to the Dev repo. It's only here to make syncing between repos easier (can do copy/paste without unpicking individual files).

scrdest commented on code/modules/urist/GOAI/__ss13/animations.dm:
DEVJUNK: Feel free to ignore this. This is a port of SS13 code to the Dev repo. It's only here to make syncing between repos easier (can do copy/paste without unpicking individual files).

scrdest commented on code/modules/urist/GOAI/__ss13/atom_stubs.dm:
DEVJUNK: Feel free to ignore this. This is a port of SS13 code to the Dev repo. It's only here to make syncing between repos easier (can do copy/paste without unpicking individual files).

scrdest commented on code/modules/urist/GOAI/__ss13/mob_stubs.dm:
DEVJUNK: Feel free to ignore this. This is a port of SS13 code to the Dev repo. It's only here to make syncing between repos easier (can do copy/paste without unpicking individual files).

scrdest commented on code/modules/urist/GOAI/__ss13/obj_stubs.dm:
DEVJUNK: Feel free to ignore this. This is a port of SS13 code to the Dev repo. It's only here to make syncing between repos easier (can do copy/paste without unpicking individual files).

scrdest commented on code/modules/urist/GOAI/__ss13/transform.dm:
DEVJUNK: Feel free to ignore this. This is a port of SS13 code to the Dev repo. It's only here to make syncing between repos easier (can do copy/paste without unpicking individual files).

Commit:
chore: Removing old stuff. Replaced duplicate goai_data dirs with compile-flag-conditional locations.

Updating is now a fully 'drag-n-drop GOAI to GOAI dir' flow.

Commit:
chore: Squashing lints, methodically.

Commit:
chore: Code quality fixes. Or adjustments, where I know for sure it's necessary and fine.

Commit:
chore: More of the same

scrdest commented on test/check-paths.sh:
```diff
@@ -30,21 +30,21 @@ exactly 2 "/datum text paths" '"/datum'
-exactly 2 "<< uses" '(?<!<)<<(?!<)' -P
-exactly 2 ">> uses" '(?<!>)>>(?!>)' -P
+exactly 82 "to_world_log uses" '\sto_world_log\('
+exactly 1 "world<< uses" 'world<<|world[[:space:]]<<'
+exactly 1 "world.log<< uses" 'world.log<<|world.log[[:space:]]<<'
```
These both had to be bumped to account for a dev-only copy of the macros file.

scrdest commented on test/check-paths.sh:
```diff
@@ -30,21 +30,21 @@ exactly 2 "/datum text paths" '"/datum'
+exactly 82 "to_world_log uses" '\sto_world_log\('
+exactly 1 "world<< uses" 'world<<|world[[:space:]]<<'
+exactly 1 "world.log<< uses" 'world.log<<|world.log[[:space:]]<<'
+exactly 14 "<< uses" '(?<!<)<<(?!<)' -P
+exactly 3 ">> uses" '(?<!>)>>(?!>)' -P
```
Similar story here: all of the extra uses are just a copy of the macros for use in the library version. Theoretically these could just be yeeted, but that would slow development down quite a bunch.

scrdest commented on test/check-paths.sh:
```diff
@@ -30,21 +30,21 @@ exactly 2 "/datum text paths" '"/datum'
+exactly 3 ">> uses" '(?<!>)>>(?!>)' -P
 exactly 0 "incorrect indentations" '^( {4,})' -P
-exactly 26 "text2path uses" 'text2path'
+exactly 37 "text2path uses" 'text2path'
```
Indispensable, because it's used by the SerDe to lookup procs from the references embedded in the file data, and there's no real alternative to text2path available (at least as of BYOND 514).

scrdest commented on test/check-paths.sh:
```diff
@@ -30,21 +30,21 @@ exactly 2 "/datum text paths" '"/datum'
 exactly 5 "goto use" 'goto '
 exactly 1 "NOOP match" 'NOOP'
-exactly 444 "spawn uses" '^\s*spawn\s*\(\s*(-\s*)?\d*\s*\)' -P
+exactly 461 "spawn uses" '^\s*spawn\s*\(\s*(-\s*)?\d*\s*\)' -P
```
All of these are internal 'ticker-esque' code which cannot use any of the weird Bay timers.

scrdest commented on debug_start_state.json:
Can remove this. Might have been added by accident.

scrdest commented on code/modules/urist/datums/xenoarch_effects/transfer.dm:
Oops, definitely should remove this; snuck in from a previous branch lol

Commit:
Sync to GOAI commit 9dbdff49481f1ebd89bed20c22041409ce3cfe9b

feature: Added macro's Chebyshev distance (since get_dist is weird and procs are slow).
feature: Added a new SO macro - grants Action if object is in direct inventory contents (no recursion).
feature: Macro-gated drawing beams
docs: Added a Utility AI Readme
tweak: Reenabled path-following mode
tweak: Added source logging to Utility brain
fix: Fixed bad SO paths for Dev repo
fix: Corrected proc used for movement
chore: Fixed some lints/dupes
chore: Big ole cleanup of debug verbs. Gated all I could find behind proc-granting procs/verbs or removed.

Commit:
feature: Adds admin verbs to self-grant GOAI verbs (including a verb to remove them)

Commit:
fix: Lintermancy

Commit:
feature: Refactoring to account for Factions/non-mob AIs better.

The concept of a Pawn is now generic; a mob AI drives a mob/atom Pawn, a faction AI drives a faction datum Pawn, a director AI would drive an Event Pawn or sth.
Reorganized Actions - split into subdirs where the Action only makes sense for a subtype.
Fixed a ton of nullchecks to be typechecks instead.

feature: Simplified the AI ecosystem.

- Rolled CombatMob AI into normal mob AI.
- Moved generic logic into the base class
- Removed Abstract AI as a separate entity - the main determinant of a subtype will be the driven Pawn type (e.g. Event data for Director AIs, Faction for Faction AIs, Mobs for Mob AIs)

Sync to SHA-1: 40c196d23bf03d071dc4e8190de6f74777f1deb8

ggermanicus commented:
I lack the code literacy to do more than stare at this blankly, but I wouldn't mind seeing this playtested in a round.

Commit:
fix: Uses the null-safety'd var for attaching, oops
fix: Defensive programs the path-following leg of movement a bit

Commit:
chore: Fighting Travis

scrdest commented on code/modules/urist/GOAI/__goai_library_features.dm:
Context: if included, this exposes the simplified, non-SS13 variants of the API. 

Shouldn't be toggled on in SS13 code, but if I remove it now, I'd have to remove it every time I pull changes from the lib to SS13.

scrdest commented on code/modules/urist/GOAI/__goai_ss13_support.dm:
Context: if included, tells the compiler to use SS13-friendly APIs.

Basically the opposite of the library_features above.

scrdest commented on code/modules/urist/GOAI/__spaceman_dmm.dm:
Another thing retained to keep syncs straightforward; if the lib changes stabilize, can be nuked.

scrdest commented on code/modules/urist/GOAI/__ss13/_ss13stubmacros.dm:
This is only really used during library development, so don't worry about it too much.

scrdest commented on code/modules/urist/GOAI/_algorithms/gaussdist.dm:
```diff
@@ -0,0 +1,199 @@
+
+	if(isnull(spare))
+		var/s = 0
+		var/u = null
+		var/v = null
+
+		while (s == 0 || s >= 1)
+			u = (uni_dist.sample() * 2) -1
+			v = (uni_dist.sample() * 2) -1
+			s = u * u + v * v
+
+		s = (-2 * log(s) / s) ** (0.5)
+		spare = v * s
+		return _mean + _std * u * s
+
+	// implicit 'else'
+	src.spare = null
+	return _mean + _std * spare
+
+
+/distribution/gauss/proc/_boxmuller_sample(var/mean = null, var/stddev = null)
+	/* Box-Muller Transform sampling */
+	var/_mean = (isnull(mean) ? src.mean : mean)
+	var/_std = (isnull(stddev) ? src.stddev : stddev)
+
+	var/distribution/uniform/uni_dist = new()
+
+	var/uniA = uni_dist.sample()
+	var/uniB = uni_dist.sample()
+
+	var/polar_dist = _std * ((-2 * log(uniA)) ** (0.5))
+	var/polar_angle = (2 * MATH_PI * uniB)
+
+
+	var/result = (polar_dist * sin(polar_angle) * _mean)
+	/* NOTE:
+	   With cos(), we could generate an extra sample out of this,
+	   but that would mean an extra tuple-handling step.
+	*/
+	return result
```
Context: those are two alternative methods for generating Gaussians out of plain rand()

scrdest commented on code/modules/urist/GOAI/_algorithms/gaussdist.dm:
```diff
@@ -0,0 +1,199 @@
+# define GAMMADIST_MAX_TRIES 1000
+
+/distribution/gamma
+	/* Gamma distribution sampler, after Marsaglia and Tsang [2000]
+	// "A Simple Method for Generating Gamma Variables"
+	*/
```
🤓

scrdest commented on code/modules/urist/GOAI/_algorithms/linear_algebra.dm:
Context: needed raycasts for things like aiming.

scrdest commented on code/modules/urist/GOAI/_algorithms/statistics.dm:
Context: this is used by some AI Considerations, e.g. "average health of visible enemies"

scrdest commented on code/modules/urist/GOAI/_algorithms/testing/linreg_debug.dm:
Debug code, doesn't really matter, excluded.

scrdest commented on code/modules/urist/GOAI/_algorithms/testing/stats_debug.dm:
Debug code, doesn't really matter, excluded.

scrdest commented on code/modules/urist/GOAI/_algorithms/testing/turfchunk_debug.dm:
Debug code, doesn't really matter, excluded.

scrdest commented on code/modules/urist/GOAI/_algorithms/trigonometry.dm:
This is only used by library compiles, it's a dupe of SS13 trig.

scrdest commented on code/modules/urist/GOAI/_datastructures/action_tracker.dm:
Context: this is the heart of action-handling for AI. 

The object defined here wraps all tasks to let us track if they are done, how long they were running, and let us store task-local variables (as assoc key/vals).

scrdest commented on code/modules/urist/GOAI/_datastructures/dict.dm:
Way back I thought this was a good idea.

It's not, will eventually replace all of these with pure assocs, but not yet.

scrdest commented on code/modules/urist/GOAI/_datastructures/lists.dm:
```diff
@@ -0,0 +1,139 @@
+
+
+
+/proc/upsert(var/list/oldassoc, var/list/newassoc)
```
This and some others are unused, will nuke. Holdover from veeeeery long time ago.

scrdest commented on code/modules/urist/GOAI/_datastructures/path_tracker.dm:
Helper datastructure for pathing code.

scrdest commented on code/modules/urist/GOAI/_datastructures/tuple.dm:
Context: needed to be able to put stuff in Priority Queues without writing custom comparison procs for every usecase.

We need THAT in turn to be able to efficiently do evaluate-options-then-pick-N-top type patterns, which are super common in AI logic.

Commit:
chore: Reverts some devcode

Commit:
feature: Fast-forwarding library version to Dec 2024

(from late Apr 2024 - massive! copypasta of commits below)

* feature: Simplified the AI ecosystem.
* Rolled CombatMob AI into normal mob AI.
* Moved generic logic into the base class
* Removed Abstract AI as a separate entity - the main determinant of a subtype will be the driven Pawn type (e.g. Event data for Director AIs, Faction for Faction AIs, Mobs for Mob AIs)
* feature: Made FactionData datum into a SmartObject. Its paths can be initialized by the Faction AI for convenience. First Faction JSON spec. Diplomancy.
* feature: Core trading logic! Can accept and make offers!
* feature: A ton of supporting building blocks (Considerations, CFs, JSON serde stuff)
* refactor: Split brain.dm into modular pieces and moved concrete brain subclasses under a shared path
* feature: Full capability to handle trading both 'actively' and 'passively'.
* feature: Need and need weight generation from JSON data.
* refactor: Move and split Utility brain to the concrete brains location.
* refactor: Remove Motives API in favor of the Needs API which it duplicated. Big projects are fun like that.
* refactor: Brain -> AI refactor

Moved the actual AI logic (evaluating and selecting actions + economy) into the AI datum.
These had previously been in the Brain datum, because the AI logic got datum'd into a Brain first, the AI class came into existence later.
This led to an unintuitive API where the Brain had a core loop logic proc, but it was getting called from the AI datum's actual core loop.
The intent behind the current API is that Brains are essentially an auxiliary AI data structure (an enriched Blackboard System, basically).
The first part is a soft launch - the APIs got ported into separate modules, the legacy APIs are still around but unincluded

feature: Basic contract lifecycle and fulfillment - currently only implemented abstractly.
feature: Asset DB (in memory of course)
feature: Tracking offers and contracts, w/ CF/Consideration integrations.
feature: Global ID system, to unify all the in-memory DB lookups
feature: Full (if instant) impls for trading, uni- or bilaterally fulfilling the deal (if possible).
feature: An API to query how much of the goods we still need to add to the Escrow to fulfill the deal.
refactor: Reworked the lifecycle flags to two separate bitflags - one for 'core' stuff, the other for 'has progressed' flags.
fix: Fixed bad AbortPlan() behaviour - would not advance to next check.
feature: An API to pause processing of Actions (if we know there's no point in an expensive proc call).
feature: Actual faction-to-faction trading!
fix: Less silly pricing
feature: Deferred pawn init for AIs (used to customize faction spawns for ^^^)
feature: Lookup-based needs (currently money)
feature: (1) Saner trade prices.
feature: (2) Binary Heap-based priority queue

(1) Last unit prices are stored in Yet Another Global Array and seeded from a JSON.
The trade AI incorporates this into their guesstimates where the last price is better than their guess, adjusting the pricing.
This stabilizes the price offers somewhat, as long as the AI needs are not extreme - a desperate AI will still make desperate deals.
Got a basic two-faction setup going! Still needs production/consumption implemented.

(2) This fixes a LOOOOOOONG-standing bad implementation of PQs in SS13.
The original implementation is O(N) worst-case on insertion due to swapping all tail items to the next slot.
The PQ is defined as an abstract class with a legacy OrderedList impl and a new BinHeap impl, with a define to switch between them.
The legacy implementation uses a strange convention that had been retained for now in both classes for integration's sake.

feature: Working trading! Details below...
feature: Production/Consumption system - a ticker that simulates processing of Assets based on 'recipes'.
feature: Asset-Needs system - a ticker that updates Needs based on the amounts of affecting Assets (e.g. food for Hunger, money for Wealth).
feature: API to set Need updates batched to avoid repeated proc-calls.
feature: A DB of 'objective' Need weights to use as a reference point
tweak: Sell offers now consider their own Utility AND the 'objective' Utility for pricing and take the maximum of the two prices.
tweak: Sell offers now have 'friction' in the form of minimum money volume required to even bother (hardcoded at 100$ ATM).
refactor: Removed the special-case fetcher for Wealth - no longer needed due to A-N system.
refactor: Moved the Attachments variable to be common across all datums.
feature: Three-way trading case - map and definitions
feature: Added effort cost to sales - reduces low-volume/low-price busywork trades
feature: ContextArg for skipping needs in ctxfetcher_get_needs()
feature: EscrowPut toggle param - require full sum upfront or allow incremental ringfencing
fix: Bad above/below ctxfetcher_get_needs (didn't know the Need max ceiling might need to be flexible)
fix: Forgot to commit the economy_asset_transforms.json :3
tweak: Added world.time timestamps to certain logs (to facilitate log visualizations mainly)
feature: 5-way Faction Trade Scenario! (miners, hi-tech farmer, lower-tech farmer, pure big refinery, boutique mining/refining op)
refactor: Asset Transforms is now a list[map[...]] instead of map[asset, map[...]]; added a key for comments
refactor: Reworked testmaps for factions - maps are now tiny and can be done additively (need to split it to map per spawner)
feature: New faction definitions (FoodCo - small but techy food producer; Hephaestus - does both mining and refinement, but only small-scale)
feature: New Consideration - can we fulfill the Contract? Should reduce Actions wasted on retrying fulfillment attempts
balance: Fast Buy offers only happen some of the time (random, weighted by need being very low)
balance: Better values on Asset Transforms & CommodityDB for a more dynamic economy
feature: Working complex economy case!
feature: Asset-needs system now has a sidecar purpose of tallying up the totals of all resources in existence
feature: Asset-needs system now also emits faction assets to logs to facilitate graphing
feature: Allowed the specification of normalization basis to Need fetchers - can use '1' to denormalize to raw values.
balance: Buy offers clamped to never request more than 75% of global resource total
fix: Split worker regen into 3 steps (hunger/fed/usable) to prevent weird explosive growth bugs
tweak: Reordered callsig checks in CIs to be able to log the source requester
tweak: General value tweaks to get the economy in a better place
feature: Added tracking of asset deltas per most recent tick.
tweak: Workers consume more food (1u/worker), but factions start with more food. Tweaked farm productivity.
fix: Workers no longer proliferate explosively. Removed the now-redundant fed worker stage.
fix: Removed buying fractional units (missed a round to integers)

Glloyd commented:
dreamchecker is very angry at this PR

scrdest commented:
> dreamchecker is very angry at this PR

Yeah, I'll beat it down. 

The fast-forward by half a year brought a lot of, um, _research-grade code_ into this, but I streamlined the Brain architecture significantly, so it was worth it. Plus, while it's probably best excluded for now, a real trade system prototype's in there.

Commit:
chore: Big warning cleanup pass (two left, but for good reason)
feature: Extracted trading logs to a macro to toggle them off with an include/debug module switch.

Commit:
chore: Squashing lints

Commit:
chore: More lints

Commit:
feature: Suppresses noncritical GOAI WIP warns unless a compile option is toggled.

CONTEXT: I've gotten in the habit of using warn directives as nice markers of things to finish/clean up as the compiler will pester me to handle them and give me a link directly to code. The warns covered by this change are exclusively WIP stuff that doesn't break anything. Keeps CI happy but provides an easy way to reenable them.

Commit:
fix: Sure would have been nice IF I ACTUALLY UNINCLUDED THE WIPs

Commit:
chore: Fighting lints more. Disabled the indentation awk - it generates false positives

Commit:
chore: The grand conditional-macroization of GOAI lib logging.

Everything has been turned into a macro that conditionally compiles into either a to_world_log() macro or nothing, depending on the debug module settings.

Commit:
chore: Updating expectations to raise limit of to_world_logs.

This is legit because all the new instances are macro definitions.

Commit:
feature: Reloading AIs/cache-busting loading JSONs
fix: Fixed dev-hijacking the UI rendering most clicks useless

Commit:
feature: Big ole optimization/fixes pass.
feature: AI LODs system - AIs, Actions and Senses are gated behind a min & max LOD value.
feature: Dynamic LOD checkers - LODs are set by an arbitrary unary function.
feature: Basic LOD proc - skip mob AIs with non-CONSCIOUS pawns.
feature: Factions relationship DB - GOAI mobs with different factions can now be friendly. Data-driven definition in a single JSON!
fix: Restored classic mob AI behavior - different factions start hostile.
tweak: Gated all the debug stuff like beam gizmos behind the include properly.
fix: Queues shouldn't runtime when empty.

Commit:
chore: Linterrrrr

Commit:
chore: Missed an include, dammit

Commit:
Squads, it's a Christmas miracle!

feature: Squads! Abstracts groups of AI, can 'collapse' units for cheaper, low-LOD processing. SmartObject, supports 'ranks/seniority', can be ordered independently.
feature: Added Squad spawners for the debug mobs. Spawns a group of mobs AND wires them up to AI Squad.
feature: Added a LOD proc that ignores nullspaced atoms' AIs - meant to be used for undeployed Squad members.
feature: GOAI Pathfinder proc now accepts a list of avoid turfs. Known occupied turfs are avoided automagically.
feature: Added more granular logging macros for Consideration eval debugging - log total + log on rejection score.
feature: Added params to the Turfs In View ContextFetcher to allow for pre-filtering to reduce pointless Consideration checks.
fix: Fixed bad Relationship logic - friendlies are friendly again.
fix: Fixed debug guns not drawing shots properly.
fix: Fixed missing LOD params on initializing some Senses.
fix: Patched up some broken SO JSONs.
refactor: Moved Faction on dummy mobs to base mob/ class (from mob/living)
refactor: Macro-ified checks for global registries to reduce maintenance load.
refactor: Moved the brain Hivemind feature to a separate module.
refactor: Made the brain Actions module more actually modular.

scrdest commented:
For the second day of Christmas, ERPspider Santa has brought the gift of Squads. 🎅🎁 

And yeah, I know, adding more stuff onto the review pile. I was going to keep it for later, but in the process of working on it, I discovered I, um, may have borked up the relationships system while porting simplemobs to legacy SS13 behavior (i.e. hostile by default).

Commit:
chore: CI fixes

Commit:
fix: Bad iterator. Unironically, thanks CI!

Glloyd commented:
i really want to just run a playtest of this beast and see how it works in practice, how feasible is that with where the system is at right now?

scrdest commented:
> i really want to just run a playtest of this beast and see how it works in practice, how feasible is that with where the system is at right now?

🙌 Thank you for getting back on this! 🙌 

Should be fine, I've tested this enough that random runtimes shouldn't be a problem unless there's some particularly weird edge-cases in a multiplayer settings, but I doubt it. The main concern is that while the AI engine is there, the AI behaviors themselves will not be super impressive as of this PR - 'cause MVP.

The system itself is designed to be a very clean plug-in/plug-out module; currently nothing runs on GOAI by default - you have to either manually proc-call on an existing generic atom to have GOAI take it over, or use a spawner that more or less does the same for you on the spawned entity. It's also LOD'd and lazy-initialized all over the place, so you don't pay for the AI stuff that is not currently used.

To be extra-safe, if you could give me a rough idea on when you wanted to playtest it, I can give it one last pre-flight check on my end beforehand to catch anything stupidly obvious I may have forgotten about, since I hadn't really touched this at all in like a month+.

Commit:
fix: Patched up melee for humanoid agent AIs. No more Combat Hugs, Punch uses pistol-whips instead of point-blank shots.
tweak: Adjusted priorities to make humanoid AIs more aggressive and less prone to taking cover all the time.

Commit:
tweak: More null-safety for AStar

Commit:
chore: Post-rebase fixes

Commit:
Some last minute fixes, because that's always a good idea.

feature: Added 3D version of Chebyshev distance
feature: Added a WIP move-and-melee attack (disabled due to bugginess currently)
feature: Gave GOAI humanoid spawner mobs a random pistol to show off the capability
feature: Optional fuzzing of pathfinds
feature: Added pursuit for recently-seen enemies
fix: Fixed up some bad multi-z pathfinding cases
fix: Removed a debug verb I left in accidentally

Commit:
fix: Temporarily disables hand-switching - causes a runtime in Baycode

scrdest commented:
Aaaaand another rebase. Need to fix up the stupid paths check. And disable the var/ checks in DreamChecker, because they dumb.

Commit:
fix: Kills the stupid no-var-in-arguments check dead. Explicit > implicit, dammit!
fix: Restores/merges check-paths.sh with latest.

Commit:
fix: No! Bad Bay!

Commit:
feature: Gizmos - debug visuals, but way cheaper thanks to render_target trickery. Currently enabled by a GOAI define.
feature: GOAIs actively update their pathing location if tracked thing moves, using their Actual Eyes (TM).
refactor: The should_pop arg for read_var removed - this was a bad idea and should never be allowed for Considerations.
fix: GOAI movement has been substantially dejanked, because of the above (and another, memory-related stupidity of mine) affecting it before.
chore: Reconciled the code with linter standards.

Commit:
fix: Some more linting issues

Commit:
fix: Nukes an unused module (holdover from renames)

Commit:
fix: Sigh... lints.

Commit:
fix: Oops duplicate hierarchy instances

Commit:
TL;DR - improved AI target-tracking significantly. Between this and the LKP changes, it's getting pretty scary - at least once I fix opening doors.

feature: Improved vision - uses LOS raycast in addition to radius for more flexibility and range.
feature: New(ish) charge-in melee Action. Old Punch only works if already adjacent, so a bit more defensive.
feature: Raytrace can now pass a maxdist that functions as an early exit.
feature: TurfDensityRaytrace can now early-exit on hitting the target.
tweak: Pathservice tries to make a path without doors/tables first, if possible.
fix: Multiple stupid oversights in raytrace code that made it not work quite as it should.
fix: Fixed Mob Commanders' InitRelationship. It's been very stupidly busted, did I code that drunk?
fix: Non-SS13 Melee was not returning anything, causing it to be perceived as failed.
fix: Patched up InitRelations to also happen on attach. Bit hacky.
fix: Updated super-stale paths in some old test maps.
refactor: Moved gizmos.dm to a higher-priority file to make it available in algos and such

Commit:
chore: Updated all the (relevant) proc refs to use proc-ref macros.

Commit:
chore: 97->98 logs -_-

Commit:
feature/fix: GOAI movement can now correctly bump things like autodoors (to open them) again. Finally.
feature: A new memory, HaltManeouverActive, can be used to tell AI to pause movement. This can be used for actions where the mobs should remain still.
feature: ctxfetcher_in_pawn_turf now now exposes a new arg, 'require_filtered_type'. If TRUE, does not create contexts if the passed filter has no matches. Prevents nulls in Considerations for tableflips.
tweak: Broadly reduced pathing penalties for obstacles to make them not dominate distance cost.
tweak: Lazified allocs for list in movement code. Less heap is more good.
fix: Patched vision gizmos to (lazy-)init gizmos - more expensive, but safety against runtimes.
fix: GOAI tableflips report success correctly

Commit:
fix: Updating test expectations

scrdest commented:
Rebased to latest and fixed doorbumping, so AIs are no longer foiled by simple airlocks.

Commit:
chore: Cleaned up unused vars by restoring stubs.

I did a dumb while debugging, this is only used in libcode so couldn't have caused the issue.
